rfcs/20210129-optimize-pd-creation/README.md

```diff
@@ -0,0 +1,262 @@
+# Optimization of Primitive Descriptor Creation
+
+## Motivation
+
+There are three majors steps to create a oneDNN primitive:
+1. Create operation descriptor
+2. Create primitive descriptor
+3. Create primitive
+
+In frameworks, all the three steps are performed every time when a user executes
+an operation. The last two steps are expensive hence they can significantly
+affect performance of the operation.
+
+oneDNN recently introduced a cache for primitives to significantly reduce primitive
+creation time - the most expensive step. The second most expensive step is creating
+the primitive descriptor. The cost may very from several microsecond to hundreds
+of microseconds depending on number of the primitive implementations and time of
+primitive descriptor initialization.
+Such overhead is significant because the execution time can also fall in this range.
+This is the motivation to start optimizing the primitive descriptor creation step.
+
+
+## Proposal
+
+Introducing one more cache for primitive descriptors doesn't look appealing for
+the following reasons:
+1. Bad user experience because user has to manage two caches
+2. Maintenance cost (maintain two multithreaded caches)
+3. Increases memory consumption
+4. New API to control one more cache
+
+The proposal is to implement an internal mechanism that would significantly
+reduce time of primitive descriptor creation but do not expose the mechanism to user.
+
+The idea is to use existing cache for primitives to implement the mechanism
+because the cache already contains all the abstractions that are created during those
+3 steps. The `primitive_t` abstraction contains `primitive_desc_t` which contains
+`op_desc_t` and `primitive_attr_t`.
+According to the primitive cache design there was added an ability to share
+the `primitive_desc_t` abstraction with `shared_ptr` therefore we can simply use it
+to create a new `primitive_desc_t` with minimum overhead.
+
+Since primitive descriptor and primitive are created for each execution of an
```

Comment:
I would suggest to put this block at the beginning of this section, since it describes the motivation for pursuing this option. Or it can be dropped altogether.

Reply:
The idea about reusing primitive cache is mentioned in the previous paragraph first time hence this sentence seems to be in the right place.